data_structures/queue/queue_on_list.py

```diff
@@ -1,52 +1,77 @@
+
+T = TypeVar("T")
+
 
 class Queue:
-    def __init__(self):
-        self.entries = []
-        self.length = 0
-        self.front = 0
-
-    def __str__(self):
-        printed = "<" + str(self.entries)[1:-1] + ">"
-        return printed
-
-    """Enqueues {@code item}
-    @param item
-        item to enqueue"""
-
-    def put(self, item):
-        self.entries.append(item)
-        self.length = self.length + 1
-
-    """Dequeues {@code item}
-    @requirement: |self.length| > 0
-    @return dequeued
-        item that was dequeued"""
-
-    def get(self):
-        self.length = self.length - 1
-        dequeued = self.entries[self.front]
-        # self.front-=1
-        # self.entries = self.entries[self.front:]
-        self.entries = self.entries[1:]
-        return dequeued
-
-    """Rotates the queue {@code rotation} times
-    @param rotation
-        number of times to rotate queue"""
-
-    def rotate(self, rotation):
-        for i in range(rotation):
-            self.put(self.get())
-
-    """Enqueues {@code item}
-    @return item at front of self.entries"""
-
-    def get_front(self):
-        return self.entries[0]
-
-    """Returns the length of this.entries"""
-
-    def size(self):
-        return self.length
+    """
+    >>> queue = Queue()
+    >>> queue.peek()
+    Traceback (most recent call last):
+    ...
+    IndexError: The queue is empty
+    >>> queue.size == 0
+    True
+    >>> queue.is_empty()
+    True
+    >>> queue.enqueue(2)
+    >>> queue.peek()
+    2
+    >>> for i in range(3, 10):
+    ...     queue.enqueue(i)
+    >>> queue.dequeue()
+    2
+    >>> queue.dequeue()
+    3
+    >>> queue.rotate(2)
+    >>> queue.dequeue()
+    6
+    """
+
+    def __init__(self) -> None:
+        self._queue: list[T] = []
+
+    def __repr__(self) -> str:
+        return f"{self._queue}"
+
+    def enqueue(self, item: T) -> None:
+        """Add an item to the queue"""
+        self._queue.append(item)
+
+    def dequeue(self) -> Union[T, NoReturn]:
+        """Remove and return the first item in the queue.
+
+        :raises IndexError if queue is empty
+        """
+
+        if not self.is_empty():
+            return self._queue.pop(0)
+
+        raise IndexError("The queue is empty")
+
+    def peek(self) -> Union[T, NoReturn]:
+        """
+        Return the first item in the queue without removing it from the queue
+
+        :return: First item in the queue.
+        :raises IndexError if queue is empty
+        """
+
+        if not self.is_empty():
+            return self._queue[0]
+
+        raise IndexError("The queue is empty")
+
+    @property
+    def size(self) -> int:
+        return len(self._queue)
+
+    def is_empty(self) -> bool:
+        return self.size == 0
+
+    def rotate(self, number_of_rotations: int) -> None:
+        """Rotate a queue by `number_of_rotation` times"""
+        for _ in range(number_of_rotations):
+            self.enqueue(self.dequeue())

```

Comment:
Can be executed more efficiently -
by combining it into a one-liner,
or using `repeat` from `itertools`.

Reply:
You are right. I didn't go with the one-liner for readability sake, but I love your idea about itertools.

Thank you.